sound/soc/sof/ipc3-control.c

```diff
@@ -647,7 +659,7 @@ static int sof_ipc3_widget_kcontrol_setup(struct snd_sof_dev *sdev,
 	list_for_each_entry(scontrol, &sdev->kcontrol_list, list)
 		if (scontrol->comp_id == swidget->comp_id) {
 			/* set kcontrol data in DSP */
-			ret = sof_ipc3_set_get_kcontrol_data(scontrol, true);
+			ret = sof_ipc3_set_get_kcontrol_data(scontrol, true, false);
```

Comment:
By definition the `widget_kcontrol_setup` op is called from `sof_widget_setup()` which is the locked, but yes, we could try.

Reply:
@plbossart, if you insist I can add
```c
WARN_ON_ONCE(!mutex_is_locked(&swidget->setup_mutex));
```
here and the ipc4 callback as well, but it will never catch anything as this is always happening at widget setup.

Reply:
I was thinking lockdep_assert_held(&swidget->setup_mutex);

Reply:
@plbossart @ujfalusi I think what we could do instead is split the sof_ipc3_set_get_kcontrol_data() into unlock and regular functions and call the unlock version from sof_ipc3_widget_kcontrol_setup() and then add the lockdep_assert() here in

Reply:
> @plbossart @ujfalusi I think what we could do instead is split the sof_ipc3_set_get_kcontrol_data() into unlock and regular functions and call the unlock version from sof_ipc3_widget_kcontrol_setup() and then add the lockdep_assert() here in

@ranj063, I did tried to split the *-control.c functions as well but it turned out to be not so pretty. We have common code to find the swidget and it will create a web of calls which is hard to follow. The additional flag parameter was the cleanest way to do this.

Reply:
> I was thinking lockdep_assert_held(&swidget->setup_mutex);

I see, that could be added with the similar note that it is just a NOP as the setup is only going to happen during the setup.